### PR TITLE
Added Counting Sort Implementation

### PR DESCRIPTION
**Pull request by Leela0o5**
I added a Python implementation of the counting sort algorithm with an example. Tested and works as expected.

### CODE REVIEW

File: sorting/count_sort/count_sort.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import math
+def counting_sort(L):
+    largest = -math.inf
+    n = len(L)
+    for i in range(n):
```
Comment (FirefoxSRV) [inline sorting/count_sort/count_sort.py:5]: 1. Edge case: If the input is an empty list ( [ ] ), the code will tend to break. Handle this case !
2. Edge case: If there are negative numbers, the code can misbehave as you have not handled negative numbers as an edge case.
3. Edge case: If the range of numbers in the list is much larger than the number of elements in L, the algorithm may not be space- and time-efficient. Example: [1, 1000000].....The range size is 1000000 - 1 + 1 = 1000000. The frequency array will have 1,000,000 entries, but only two of those entries will be used. This results in high space complexity (O(range_size)), even though the time complexity is linear (O(n + range_size)).

**Review by FirefoxSRV — CHANGES_REQUESTED** [review]
Requested to handle edge cases

File: sorting/count_sort/count_sort.py (hunk 1 above)
Comment (Leela0o5) [reply sorting/count_sort/count_sort.py:10]: Hey! I implemented and handled the first 2 edge cases. please check it out. The 3rd edge case that you described is the real disadvantage of counting sort and it is suitable for small-range data like student marks which are in the range(1 -100) and are well known to us and obvious. For the 3rd edge case, we can go for an alternate sorting method.

Comment (FirefoxSRV) [reply sorting/count_sort/count_sort.py:10]: Cool...edge cases handled

Comment (FirefoxSRV) [reply sorting/count_sort/count_sort.py:10]: Can you acknowledge if this is a code from GPT ? If not, we can have a meet to discuss upon

Comment (Leela0o5) [reply sorting/count_sort/count_sort.py:10]: Hi @FirefoxSRV. For the second edge case, I referred to a research paper on the net that implemented this similar shifting logic to handle negative numbers, which I did not understand properly.
I wrote the previous code by myself and to handle edge case number 2 easily, I referred to this website
 https://yuminlee2.medium.com/counting-sort-algorithm-392560fe570e.I implemented the same logic Please do check it out.

Comment (Ashrockzzz2003) [reply sorting/count_sort/count_sort.py:10]: Awesome @Leela0o5.
If you can add the research paper link to the code on top, it'll be great!

Comment (FirefoxSRV) [reply sorting/count_sort/count_sort.py:10]: Good one @Leela0o5 !

**Review by FirefoxSRV — APPROVED** [review]

**Review by FirefoxSRV — APPROVED** [review]

File: sorting/count_sort/count_sort.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# https://www.ripublication.com/ijaer18/ijaerv13n1_28.pdf
+import math
```
Comment (Ashrockzzz2003) [inline sorting/count_sort/count_sort.py:2]: Format the comment like this for better readability and remove unused import `math`.

```suggestion
"""
Logic to handle negative numbers in counting sort was referred from the below link:
https://www.ripublication.com/ijaer18/ijaerv13n1_28.pdf
"""
```

**Review by Ashrockzzz2003 — CHANGES_REQUESTED** [review]
Just one last thing Leela. Fix it and we'll merge it!

File: sorting/count_sort/count_sort.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+https://www.ripublication.com/ijaer18/ijaerv13n1_28.pdf
+"""
+def counting_sort(L):
+    if not L:
+        return L
+    largest = max(L)
+    min_value = min(L)
+    offset = -min_value if min_value<0 else 0
+    n = len(L)
+    frequencyArray = [0]*(largest+1+offset)
```
Comment (Ashrockzzz2003) [inline sorting/count_sort/count_sort.py:12]: The comment is not aligned with the padded block. Please update formatting.

File: sorting/count_sort/count_sort.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    for i in range(n):
+     # shift of the element  by the offset which ensures negative numbers are shifted to non-negative indices in frequencyArray
+        frequencyArray[L[i]+offset] = frequencyArray[L[i]+offset]+1
+    #sorting part
```
Comment (Ashrockzzz2003) [inline sorting/count_sort/count_sort.py:16]: Usually a space is left after a comment token and the upper and lower cases are followed exactly like English. And followed by a full stop in the end.
`# Sorting part.`

Update it here as well as places below.

**Review by Ashrockzzz2003 — CHANGES_REQUESTED** [review]
Requested some changes regarding formatting of comments. Doing these will fetch you bonus points.

## Resources

https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings

File: sorting/count_sort/count_sort.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+def counting_sort(L):
+    if not L:
+        return L
+        
+    largest = max(L)
+    min_value = min(L)
+    offset = -min_value if min_value < 0 else 0
+    n = len(L)
+    
+    # Initialize the frequency array.
```
Comment (Ashrockzzz2003) [inline sorting/count_sort/count_sort.py:14]: the below line in itself is self-explanatory. No need for this comment.

**Review by Ashrockzzz2003 — CHANGES_REQUESTED** [review]
Just one last thing. Please update.

File: sorting/count_sort/count_sort.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    frequencyArray = [0]*(largest+1+offset)
+
+    # Count the frequency of each element in the array.
+    for i in range(n):
+        # Shift the element by the offset which ensures negative numbers are shifted to non-negative indices in the frequencyArray.
```
Comment (Ashrockzzz2003) [inline sorting/count_sort/count_sort.py:19]: comment line length too long. make it a two-line comment.
https://google.github.io/styleguide/pyguide.html#32-line-length

Comment (Leela0o5) [reply sorting/count_sort/count_sort.py]: wow, I learned how to comment properly.

**Review by Ashrockzzz2003 — APPROVED** [review]
LGTM
Awesome Leela.
Hope you learned something from this though it took longer to complete it!

**Pull request merged**